.github/workflows/ci-post-commit-analyzer.yml

```diff
@@ -0,0 +1,69 @@
+name: Post-Commit Static Analyzer
+
+permissions:
+  contents: read
+
+on:
+  push:
+    branches:
+      - 'release/**'
+    paths:
+      - 'llvm/**'
+      - '.github/workflows/ci-post-commit-analyzer.yml'
+  pull_request:
+    paths:
+      - '.github/workflows/ci-post-commit-analyzer.yml'
+  schedule:
+    - cron: '30 0 * * *'
+
+concurrency:
+  group: >-
+    llvm-project-${{ github.workflow }}-${{ github.event_name == 'pull_request' &&
+      ( github.event.pull_request.number || github.ref) }}
+  cancel-in-progress: ${{ startsWith(github.ref, 'refs/pull/') }}
+
+jobs:
+  post-commit-analyzer:
+    if: >-
+      github.repository_owner == 'llvm' &&
+      github.event.action != 'closed'
+    runs-on: ubuntu-22.04
+    env:
+      LLVM_VERSION: 18
+    steps:
+      - name: Checkout Source
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+      - name: Install Dependencies
+        run: |
+          sudo echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-18 main" | sudo tee -a /etc/apt/sources.list.d/llvm.list
+          wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+          sudo apt-get update
+          sudo apt-get install \
+            cmake \
+            ninja-build \
+            perl \
+            clang-tools-$LLVM_VERSION \
+            clang-$LLVM_VERSION
+
+      - name: Configure
+        run: |
+          scan-build-$LLVM_VERSION \
+              --use-c++=clang++ \
+              --use-cc=clang \
+              cmake -B build -S llvm -G Ninja \
```

Comment:
We could.  I wasn't sure how to decide which ones to enable, so I just left them all out .  If you have a suggested list of projects, I can do some benchmarks to see how long the builds will take.

Reply:
Long-term I don't see a strong need to exclude any project. As long as it's built with clang, the static analyzer may be able to figure something out.

For the initial evaluation it may be a good idea to stick to projects with coding style similar to LLVM, because they'll probably have similar true/false positive rates.

I think `clang` is definitely good to go, maybe let's start with that? (It'll also imply all the fun of having the static analyzer analyze the static analyzer.)

Reply:
Adding clang puts us over the 6 hour timeout limit for the job.  Is there some way I can optimize the checker to take less time?  Can I use ccache?  We could also just let it run for 6 hours and then upload the results when it's done.  I guess it doesn't matter much if we can't build all of clang.

Reply:
I'm guessing we could get speed-ups with PGO+ThinLTO+BOLT assuming perf-training on a good workload. That would be quite a bit of work to get going though. Might be something to look into down the road if this ends up getting used. We could split jobs like we've started doing in other places, but that would also take significant effort to implement.

I don't think I like the idea of just running until the job times out and then uploading whatever is there. I think that might make some issues pop up/go away each invocation depending upon what files it gets, decreasing the signal to noise ratio.

Reply:
Oh.

Hmm.

Yes, the static analyzer can be tweaked for performance. In fact the entire idea is that the static analyzer tries to "explore all possible execution paths" which fundamentally takes an infinite amount of time. So the only reason it terminates at all, is because it implements a bunch of "budgets", hard cutoffs to limit the exploration time. A limit on the total number of simulation steps across the entire graph of paths, a limit on the number of iterations spent in a loop, a limit on the number of nested function calls. (To be fair, it also tries to do its best within the budget, eg. it prioritizes the paths that lead to previously unexplored lines of code. So increasing those limits yields rapidly diminishing returns.)

There's the "user-friendly" "deep"/"shallow" analysis mode setting that tweaks a bunch of budgets at once to produce two "supported" configurations; the shallow mode is faster but less precise. Unfortunately I'm afraid that the shallow mode was historically fine-tuned for Objective-C so I wouldn't really recommend it for any other purposes.

I think the easiest/safest thing to do is to simply decrease `-analyzer-config max-nodes=225000` to a smaller value. That's the total number of simulation steps. It affects analysis time non-linearly because many "entry-point" functions are simulated in a finite amount of steps so they never reach the limit anyway. But either way it's the simplest option to tweak because when this budget runs out the analysis simply terminates. It will simply miss some deeper bugs but it shouldn't introduce a lot of new false positives. Every other option means that the static analyzer gets more time to explore a different part of the code so it becomes significantly more chaotic.

So, can you try
```
scan-build ... -analyzer-config max-nodes=150000
```
or
```
scan-build ... -analyzer-config max-nodes=75000
```
to see if you can fit into the 6-hour limit this way?

----

Also, well, it's sad that we have to sacrifice thoroughness in order to cover more ground. This means that we'll need to sacrifice most of the other sub-projects, unless they're so self-contained that we can set up a new job with just them.

I hope that eventually we'll either get faster machines, or figure out how to split analysis into several smaller jobs.

We should probably consider `-DLLVM_ENABLE_MODULES=1` as well. Fundamentally the static analyzer isn't affected by modules so it should be a pure win. Not by a lot: path exploration time would probably outweigh compilation time anyway. But it could buy us an hour or so.

There's also the part where the code is compiled normally during analysis, and the static analyzer is invoked in a separate process from normal compilation, so all the code actually gets compiled twice. At least up to the AST, and then it goes in different directions. The output produced by normal compilation isn't used for the analysis so this is fundamentally unnecessary and could in theory be cut. Except we still need to compile the binaries that we need for compiling the rest of the code, such as `TableGen`. So in case of LLVM this is hard to avoid.

----

> Can I use ccache?

That's an excellent question, I've somehow never even considered this. There's nothing fundamentally impossible about this: the static analyzer is fully deterministic so it'll produce the same output given the same translation unit and flags as input. But it'll probably be very annoying to marry ccache and scan-build, because scan-build comes with its own compiler wrapper. Also I'm not sure ccache would understand where our output is, because what scan-build passes to clang as `-o` is a directory.

So let me try and see if there's a straightforward solution, but if not, I probably won't have time to implement a proper solution.

It may also be a good idea to drop `scan-build` entirely and implement the whole thing with a tiny custom `CMAKE_CXX_COMPILER_LAUNCHER`. That's the whole point of `scan-build` anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append `--analyze` to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

Reply:
> It may also be a good idea to drop `scan-build` entirely and implement the whole thing with a tiny custom `CMAKE_CXX_COMPILER_LAUNCHER`. That's the whole point of `scan-build` anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append `--analyze` to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

What would a custom `CMAKE_CXX_COMPILER_LAUNCHER` need to do?  Just call `clang` to compile the object and then `clang --analyze` to analyze it?

Reply:
> It may also be a good idea to drop scan-build entirely and implement the whole thing with a tiny custom CMAKE_CXX_COMPILER_LAUNCHER. That's the whole point of scan-build anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append --analyze to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

If done this way, one would need to do some extra work to wrap up the analyze outputs and gather it all into a common report, like scan-build does, right? Outside of the context of this PR, I'd be interested in knowing how to do this (I'm not sure if there's any good docs that show examples on how to do this)?

Reply:
`-analyzer-config max-nodes=75000` helped get the builds under the time out.  I've also switch to using a more optimized compiler and added ccache for the compiler that actually compiles the code, but not for the compiler that does the static analysis.

Reply:
> What would a custom CMAKE_CXX_COMPILER_LAUNCHER need to do? Just call clang to compile the object and then clang --analyze to analyze it?

Yes, something like this:
```sh
#!/bin/sh
"$@"
"$@" --analyze --analyzer-output html -o $OUTPUT_DIR \
    -Xclang -analyzer-config -Xclang max-nodes=150000
```

> Also, would it be better to just compile the project normally and then use clang-tidy to run the analyzer using the compilation database?

This probably wouldn't be faster because you'll still need to regularly re-generate the compilation database. On a buildbot this implies "every time". Also clang-tidy doesn't know how to produce HTML output, and static analyzer reports are incredibly hard to read without a GUI that overlays them on top of your source code (even though technically they the CLI gives you all the same information).

But the upside would be that you'll get clang-tidy checks too, without the *third* recompilation. I wanted to properly "marry" them for a long time so that this wasn't a choice you had to make, but never got to finishing it :(

> If done this way, one would need to do some extra work to wrap up the analyze outputs and gather it all into a common report, like scan-build does, right?

Yes, you can bring `scan-build` back for this last part:
```sh
scan-build --generate-index-only $OUTPUT_DIR
```
Btw, this works because HTML reports are machine-readable due to nice magic comments:
```html
  147 <!-- BUGDESC Dereference of null pointer (loaded from variable 'x') -->
  148
  149 <!-- BUGTYPE Dereference of null pointer -->
  150
  151 <!-- BUGCATEGORY Logic error -->
  152
  153 <!-- BUGFILE /Users/artemdergachev/test/test.c -->
  154
  155 <!-- FILENAME test.c -->
  156
  157 <!-- FUNCTIONNAME foo -->
  158
  159 <!-- ISSUEHASHCONTENTOFLINEINCONTEXT 9c4d6241eb00bff2310e3e3391b46023 -->
  160
  161 <!-- BUGLINE 83 -->
  162
  163 <!-- BUGCOLUMN 10 -->
  164
  165 <!-- BUGPATHLENGTH 2 -->
  166
  167 <!-- BUGMETAEND -->
  168 <!-- REPORTHEADER -->
```
So `scan-build` doesn't need to talk to clang or observe analysis in real time in order to rebuild the index.

And if you like you can do a lot of custom post-processing on top of it, like filter out reports by some of these categories, or separate them into different directories to build a separate index file for each project.

Reply:
> Yes, something like this:
> 
> ```shell
> #!/bin/sh
> "$@"
> "$@" --analyze --analyzer-output html -o "$OUTPUT_DIR" \
>      -Xclang -analyzer-config -Xclang max-nodes=150000
> ```

OK, let me try that.